src/React/Basic.purs

```diff
@@ -35,3 +38,50 @@ react
      }
   -> ReactComponent props
 react { initialState, render } = react_ { initialState, render: mkFn3 render }
+
+
+foreign import reactStatefulComponent_
+  :: forall props state action eff
+   . { initialState :: props -> state
+     , receiveProps :: Fn2 props state state
+     , reducer :: EffFn4 (react :: ReactFX | eff) action state (action -> Eff (react :: ReactFX | eff) Unit) (state -> Eff (react :: ReactFX | eff) Unit) Unit
+     , render :: Fn3 props state (action -> Eff (react :: ReactFX | eff) Unit) JSX
+     }
+  -> ReactComponent props
+
+-- | TODO: docs
+statefulComponent'
+  :: forall props state action eff
+   . { initialState :: props -> state
+     , receiveProps :: props -> state -> state
+     , reducer :: action -> state -> (action -> Eff (react :: ReactFX | eff) Unit) -> Eff (react :: ReactFX | eff) state
+     , render :: props -> state -> (action -> Eff (react :: ReactFX | eff) Unit) -> JSX
+     }
+  -> ReactComponent props
+statefulComponent' { initialState, receiveProps, reducer, render } =
+  reactStatefulComponent_
+    { initialState
+    , receiveProps: mkFn2 receiveProps
+    , reducer: mkEffFn4 reducer_
+    , render: mkFn3 render
+    }
+  where
+    reducer_ action state send callback = do
+      newState <- reducer action state send
+      callback newState
+      
+-- | TODO: docs
+statefulComponent
```

Comment:
I think we could bikeshed the name a little here, since regular components are also stateful, but that's not a big deal.

Reply:
True haha.. `reason-react` uses `reducerComponent`, but that's not very "purescripty".  It should at least feel similar to the other names (like all or none ending with "Component").

Cases:

1. "stateless" -- this probably won't exist as a "component" because simple functions give you the same thing
1. "setState" (and `receiveProps` variant) -- currently `react`
1. "reducer" (and `receiveProps` variant) -- `statefulComponent` in this PR

Maybe we can avoid duplicating those component functions for `receiveProps` the same way props are type checked but optional.

Reply:
Also, this whole PR really is the same functionality in a slightly different style.  Maybe it makes more sense to leave it as-is for the sake of simplicity.  Or, if there really is value to this style and it's functionally equivalent then maybe use it instead.

I'm going to mess with this a little more this weekend and maybe just run with the reducer style all the way for comparison sake.  Then you can close it or take some or all of it depending on how it feels.